util/reggen/fpv_csr.sv.tpl

```diff
@@ -87,36 +87,54 @@ module ${block.name}_csr_assert_fpv import tlul_pkg::*; import ${block.name}_reg
   property wr_P(bit [${addr_msb}:0] addr, bit [TL_DW-1:0] act_data, bit regen,
                 bit [TL_DW-1:0] mask, int lsb);
     logic [TL_DW-1:0] id, exp_data;
-    (device_wr_S(addr), id = h2d.a_source, exp_data = h2d.a_data & a_mask_bit & mask) ##1
+    (device_wr_S(addr) and regen, id = h2d.a_source, exp_data = h2d.a_data & a_mask_bit & mask) ##1
         first_match(##[0:$] d2h.d_valid && d2h.d_source == id) |->
-        (d2h.d_error || (act_data << lsb) == exp_data || !regen);
+        (d2h.d_error || (act_data << lsb) == exp_data);
+  endproperty
+
+  // this property is to check when regen is set to 0, the write data should not be updated
+  property wr_regen_P(bit [${addr_msb}:0] addr, bit [TL_DW-1:0] act_data, bit regen,
+                bit [TL_DW-1:0] mask, int lsb);
+    logic [TL_DW-1:0] id;
+    (device_wr_S(addr) and !regen, id = h2d.a_source) ##1
+        first_match(##[0:$] d2h.d_valid && d2h.d_source == id) |->
+        (d2h.d_error || $stable(act_data << lsb));
```

Comment:
$stable only check 2 cycle periods, looks like it will cover more if we compare the value at the end of data phase with value at address phase

Reply:
Thanks Weicai for your suggestion. I thought about this but i think there might a corner case like this:
If reg `A` is gated by enable register `regen`. (This regen is internal HW controlled)
1. Write `A`, addr_phase `regen` is 1. (non-blocking)
2. Write `A` again, addr_phase `regen` is 0. (non-blocking)
3. First write `A` data_phase, `A` value is updated.
4. Second write `A` data_phase, I think there could be an assertion failure, because the reg `A` value might be updated by the first write.

Reply:
Actually I think you are right. we should be able to support this in assertion

Reply:
Great, the change LGTM